x-pack/plugins/enterprise_search/public/applications/app_search/components/document_creation/document_creation_logic.ts

```diff
@@ -74,5 +99,122 @@ export const DocumentCreationLogic = kea<
+      },
+    ],
+  }),
+  listeners: ({ values, actions }) => ({
+    onSubmitFile: async () => {
+      const { fileInput } = values;
+
+      if (!fileInput) {
+        return actions.setErrors([DOCUMENT_CREATION_ERRORS.NO_FILE]);
+      }
+      try {
+        const textInput = await readUploadedFileAsText(fileInput);
+        actions.setTextInput(textInput);
+        actions.onSubmitJson();
+      } catch {
+        actions.setErrors([DOCUMENT_CREATION_ERRORS.NO_VALID_FILE]);
+      }
+    },
+    onSubmitJson: () => {
+      const { textInput } = values;
+
+      const MAX_UPLOAD_BYTES = 50 * 1000000; // 50 MB
+      if (Buffer.byteLength(textInput) > MAX_UPLOAD_BYTES) {
+        actions.setWarnings([DOCUMENT_CREATION_WARNINGS.LARGE_FILE]);
+      }
+
+      let documents;
+      try {
+        documents = JSON.parse(textInput);
+      } catch (error) {
+        return actions.setErrors([error.message]);
+      }
+
+      if (Array.isArray(documents)) {
+        actions.uploadDocuments({ documents });
+      } else if (isPlainObject(documents)) {
+        actions.uploadDocuments({ documents: [documents] });
+      } else {
+        actions.setErrors([DOCUMENT_CREATION_ERRORS.NOT_VALID]);
+      }
+    },
+    uploadDocuments: async ({ documents }) => {
+      const { http } = HttpLogic.values;
+      const { engineName } = EngineLogic.values;
+
+      const CHUNK_SIZE = 100;
+      const MAX_EXAMPLES = 5;
+
+      const promises = chunk(documents, CHUNK_SIZE).map((documentsChunk) => {
+        const body = JSON.stringify({ documents: documentsChunk });
+        return http.post(`/api/app_search/engines/${engineName}/documents`, { body });
+      });
+
+      try {
+        const responses = await Promise.all(promises);
+        const summary: DocumentCreationSummary = {
+          errors: [],
+          validDocuments: { total: 0, examples: [] },
+          invalidDocuments: { total: 0, examples: [] },
+          newSchemaFields: [],
+        };
+        responses.map((response) => {
+          if (response.errors?.length > 0) {
+            summary.errors = uniq([...summary.errors, ...response.errors]);
+            return;
+          }
+          summary.validDocuments.total += response.validDocuments.total;
+          summary.invalidDocuments.total += response.invalidDocuments.total;
+          summary.validDocuments.examples = [
+            ...summary.validDocuments.examples,
+            ...response.validDocuments.examples,
+          ].slice(0, MAX_EXAMPLES);
+          summary.invalidDocuments.examples = [
+            ...summary.invalidDocuments.examples,
+            ...response.invalidDocuments.examples,
+          ].slice(0, MAX_EXAMPLES);
+          summary.newSchemaFields = uniq([...summary.newSchemaFields, ...response.newSchemaFields]);
+        });
+
+        if (summary.errors.length > 0) {
+          actions.setErrors(summary.errors);
+        } else {
+          actions.setSummary(summary);
```

Comment:
A pattern to try out in the future would be to avoid individual set methods like this and instead have "success" and "failure" actions for listeners.

https://kea.js.org/docs/guide/concepts#listeners

So instead of calling `setErrors` on failure, and `setSummary` + `setCreationStep` on success ... call `uploadDocumentsFailure` or `uploadDocumentsSuccess` respectively, and handle setting `errors`, `summary`, and `creationStep` in reducers that listen for those actions.

Reply:
I'd love to have a team session (maybe a tech talk or even just a rubber ducky) about this sometime actually - I struggle a little bit about writing things "the Kea way" because it feels super unintuitive to me. Like, I get that using reducers saves us extra `actions.something` calls, but to me having the action fns listed is really explicit and helpful from a developer POV as opposed to me having to ctrl+F all over the file to figure out what all vars/values an action happens to affect (vs. having them listed clearly in a row).

Definitely would be super helpful for me to discuss this as a team!

Reply:
Yeah, its advice I've heard more than once, not only in the Kea context but also for Redux. As far as I understand it is common wisdom. It would be interesting to hear arguments against though, for sure.

Reply:
Ya I'd def love to see a code diff of what this logic file for example would look like. The failure/success actions make sense & are fairly straightforward, what's tripping me up I think is what to replace `setWarnings` with since it feels like that goes along with `setErrors` and `setSummary`.

I think I'll leave this as-is for now since I'm not sure what the 'ideal' code should look like and I'd rather walk it through with someone who's got a clearer picture in their head than me, but I'm very very interested in doing that w/ someone next year!

Reply:
I think what's also extra weird for me for *this* specific logic file / use case is that the summary can also include "errors" (aka invalid documents) which is a failure state, so it's not 100% a success state... and it's really hard to wrap my mind around that / describe that accurately via method names other than just 'showSummary'/'setSummary'.

Reply:
> Ya I'd def love to see a code diff of what this logic file for example would look like. The failure/success actions make sense & are fairly straightforward, what's tripping me up I think is what to replace setWarnings with since it feels like that goes along with setErrors and setSummary.

Basically, we'd rip out all of the "set" prefixed "setter" actions. We'll instead just think about what events occur in our UI, and what state needs to change when those actions occur.

`uploadDocumentsFailure` and `uploadDocumentsSuccess` are two of those events. So we turn those into actions.

What state needs to change when `uploadDocumentsFailure` occurs? `errors` and `isUploading`. We handle that in reducers:

```js
errors: [
  [],
  {
    uploadDocumentsFailure: (_, { errors }) => (Array.isArray(errors) ? errors : [errors]),
  },
],
isUploading: [
  [],
  {
    uploadDocumentsFailure: () => false,
  },
],
```

What state needs to change when `uploadDocumentsSuccess` occurs? `summary`, `creationStep`, and `isUploading`. We handle that in reducers:

```js
creationStep: [
  DocumentCreationStep.AddDocuments,
  {
    uploadDocumentsSuccess: (_, { creationStep }) => creationStep,
  },
],
isUploading: [
  [],
  {
    uploadDocumentsFailure: () => false,
    uploadDocumentsSuccess: () => false,
  },
],
summary: [
  {} as DocumentCreationSummary,
  {
    uploadDocumentsSuccess: (_, { summary }) => summary,
  },
],
```

`warnings` needs to be set when `onSubmitJson` triggers. If we have access to the entire state in a reducer (not sure if we do or not, we could do it this way):

```js
warnings: [
  [],
  {
    onSubmitJson: (_, { warnings }, { textInput }) => {
      const MAX_UPLOAD_BYTES = 50 * 1000000; // 50 MB
      if (Buffer.byteLength(textInput) > MAX_UPLOAD_BYTES) {
        return [DOCUMENT_CREATION_WARNINGS.LARGE_FILE];
      }
      return [];
    }
  },
],
```

If not, we can create a new event, like `uploadDocumentInitiated`:

```js
warnings: [
  [],
  {
    uploadDocumentInitiated: (_, { warnings }) => {
      return warnings || [];
    }
  },
],
```

I see a few benefits of this approach:
1. When you use the Redux debug console, it reads like a story. It makes debugging much simpler. You see "what happened" (the event), and then "how state changed" in one singular state diff. When using the current approach, you see a bunch of disconnected "setXYZ" actions that aren't tied together in any way and you have no idea where they originated.
2. Likewise, code and tests reads easier. "what can happen", and "how does state change when that happens?"
3. If you want to listen for actions from one reducer in another reducer, it's much more intuitive to listen of "events" then setter methods. If a bunch of setters get called for a particular event, which setter do you lsiten for?

Reply:
This is super interesting and helpful, thanks! I love the 'reads like a story' analogy, that really helps me wrap my head around it better.

Super interested in the 1st `warnings` reducer example you gave - I don't think we can access other state within reducers (from my quick skimming of Kea at least), so 2nd option might be the way to go (also makes it more flexible in the event that we add more warnings that aren't specifically tied to `textInput`).

Reply:
I think I'm going to go ahead and merge this PR in since I have another test helper PR that depends on this work, but I'll see if I can't open up a new PR with your proposed changes as a follow-up and continue the discussion there.

Reply:
Oh yeah sorry, this is totally non-blocking feedback. I don't think you need to make that change now, or ever. I think there's low value in going back and rewriting things in that style. I think there can be value for new code that we write, however, which is why I opened up the conversation.

Reply:
I'm using that style now in a new logic file that I'm writing, so your value has been realized!! 😁 

One thing that I'd also like to chat about (maybe in the PR that I'll be opening shortly) is naming for actions that follow this style. I'll admit a lot of this is habit on my part, but historically for functions I try to name them with `verbSomething` (e.g., `setX`, `fetchY`, `getZ`) instead of a name that sounds like a noun - this makes it slightly easier for me to distinguish functions from variables at a glance. That breaks a bit for this style of writing logic files, so I tried out writing story actions with an `onX` prefix - e.g., `onUploadDocumentsSuccess`, `onFetchY`, and so forth.

Let me know what you think about that naming convention? I'm a little torn because `onX` var names makes me think of jQuery and callback hells for some reason but I also think it makes sense in this context.

Reply:
Haha. I think a past tense verb would be appropriate for an action name.

> uploadDocumentsSucceeded

So you're saying "what happened" in code as you write it.

Reply:
Hmm! My only thought is that could definitely be mistaken for a boolean at a glance - we've used past tense for boolean states in the standalone UI. Normally I prefer a `isX`/`hasY`/`didZ` prefix/naming to indicate bools, but we're currently not consistent on that. It also could be I'm being too prescriptive about my var/fn/bool names based on my past experiences - I've found it super helpful previously, but it's potentially not anymore since we have Typescript and Typescript should in theory be checking and documenting types for us.